File: src/main/java/com/goudourasv/http/lectures/LecturesResource.java
package com.goudourasv.http.lectures;

import com.goudourasv.domain.lectures.Lecture;
import com.goudourasv.domain.lectures.LecturesService;
import com.goudourasv.http.lectures.dto.LectureCreate;
import com.goudourasv.http.lectures.dto.LectureUpdate;
import io.smallrye.common.annotation.Blocking;

import javax.enterprise.context.ApplicationScoped;
import javax.validation.Valid;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.UriInfo;
import java.net.URI;
import java.util.List;
import java.util.UUID;


@ApplicationScoped
@Path("/lectures")
public class LecturesResource {
    private final LecturesService lecturesService;

    // @Inject not needed in Quarkus
    public LecturesResource(LecturesService lecturesService) {
        this.lecturesService = lecturesService;
    }

    @Blocking
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public List<Lecture> getLectures(@QueryParam("Course") UUID courseId) {
        try {
            List<Lecture> filteredLectures = lecturesService.getFilteredLectures(courseId);
            return filteredLectures;
        } catch (Exception ex) {
            throw new ServerErrorException("Something went wrong", Response.Status.INTERNAL_SERVER_ERROR);
        }
    }

    @Blocking
    @GET
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public Lecture getSpecificLecture(@PathParam("id") UUID lectureId) {
        Lecture lecture = lecturesService.getSpecificLecture(lectureId);
        return lecture;
    }

    @Blocking
    @POST
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public Response createLecture(@Valid LectureCreate lectureCreate, UriInfo uriInfo) {
        Lecture lecture = lecturesService.createLecture(lectureCreate);
        String path = uriInfo.getPath();
        String location = path + "/" + lecture.getId();
        return Response.created(URI.create(location)).entity(lecture).build();
    }

    @Blocking
    @DELETE
    @Path("/{id}")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public void deleteLecture(@PathParam("id") UUID id) {
        boolean deleted = lecturesService.deleteSpecificLecture(id);
        if (!deleted) {
            throw new NotFoundException("Lecture with id: " + id + "doesn't exist");
        }
    }

    @Blocking
    @PUT
    @Path("/{id}")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public Lecture updateLecture(@PathParam("id") UUID id, LectureCreate lectureCreate) {
        Lecture lecture = lecturesService.updateLecture(id, lectureCreate);
        return lecture;
    }

    @Blocking
    @PATCH
    @Path("/{id}")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public Lecture partiallyUpdateLecture(@PathParam("id") UUID id, LectureUpdate lectureUpdate) {
        Lecture lecture = lecturesService.partiallyUpdateLecture(id, lectureUpdate);
        return lecture;

    }


}

File: src/main/java/com/goudourasv/http/lectures/LecturesResource.kt
package com.goudourasv.http.lectures

import com.goudourasv.domain.lectures.Lecture
import com.goudourasv.domain.lectures.LecturesService
import com.goudourasv.http.lectures.dto.LectureCreate
import com.goudourasv.http.lectures.dto.LectureUpdate
import io.smallrye.common.annotation.Blocking
import java.net.URI
import java.util.*
import javax.enterprise.context.ApplicationScoped
import javax.validation.Valid
import javax.ws.rs.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response
import javax.ws.rs.core.UriInfo

@ApplicationScoped
@Path("/lectures")
class LecturesResource(private val lecturesService: LecturesService) {

    @Blocking
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun getLectures(@QueryParam("courseId)") courseId: UUID?): List<Lecture> {
        return lecturesService.getFilteredLectures(courseId)
    }

    @Blocking
    @GET
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun getSpecificLecture(@PathParam("id") lectureId: UUID): Lecture {
        return lecturesService.getSpecificLecture(lectureId)
    }

    @Blocking
    @POST
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun createLecture(@Valid lectureCreate: LectureCreate, uriInfo: UriInfo): Response {
        val createdLecture = lecturesService.createLecture(lectureCreate)
        val path = uriInfo.path
        val location = path + "/" + createdLecture.id
        return Response.created(URI.create(location)).entity(createdLecture).build()
    }

    @Blocking
    @DELETE
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun deleteLecture(@PathParam("id") lectureId: UUID) {
        val deleted = lecturesService.deleteSpecificLecture(lectureId)
        if (!deleted) {
            throw NotFoundException("Lecture with id: $lectureId  doesn't exist")
        }
    }

    @Blocking
    @PUT
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun updateLecture(@Valid lectureCreate: LectureCreate, id: UUID): Lecture {
        return lecturesService.replaceLecture(id, lectureCreate)
    }

    @Blocking
    @PATCH
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun partiallyUpdateLecture(lectureUpdate: LectureUpdate, id: UUID): Lecture {
        return lecturesService.partiallyUpdateLecture(id, lectureUpdate)
    }

}